📖 update releasing.md

tuminoid opened the pull request:
Update releasing.md with the up to date steps for creating CAPM3 release.

tuminoid commented:
/hold

kashifest commented:
/test-ubuntu-integration-main

kashifest commented:
/test markdownlint

tuminoid commented:
/unhold
/test-ubuntu-integration-main 
/test-centos-e2e-integration-main 

/cc @kashifest @smoshiur1237 @mboukhalfa @furkatgofurov7 PTAL

tuminoid commented:
/test-ubuntu-integration-main
/test-centos-e2e-integration-main

smoshiur1237 reviewed (approved):
It is in better shape and understandable step by step instruction. 
LGTM from my side, lets get review by others.

kashifest commented on docs/releasing.md:
```diff
@@ -1,116 +1,141 @@
 
-## Creating a release for CAPM3
+Clone out the repository:
+`git clone git@github.com:metal3-io/cluster-api-provider-metal`
```
```suggestion
`git clone git@github.com:metal3-io/cluster-api-provider-metal3`
```

tuminoid replied:
Fixed.

kashifest commented on docs/releasing.md:
```diff
@@ -1,116 +1,141 @@
+Things you should do before making a release:
 
-## Prerequisites
+- Uplift controller to use latest corresponding CAPI modules
```
```suggestion
- Uplift controller go modules to use latest corresponding CAPI modules
```

tuminoid replied:
Fixed.

kashifest commented on docs/releasing.md:
```diff
@@ -1,116 +1,141 @@
 
-### Expected artifacts
+These permissions are implicit for the org admins and repository admins.
+Release team gets their permissions via `metal3-release-team` membership. Adding
```
```suggestion
Release team gets the permissions via `metal3-release-team` membership. Adding
```

tuminoid replied:
Fixed.

kashifest commented on docs/releasing.md:
```diff
@@ -1,116 +1,141 @@
+These permissions are implicit for the org admins and repository admins.
+Release team gets their permissions via `metal3-release-team` membership. Adding
+person to the team gives him/her the necessary rights in all relevant
+repositories in the organization. Individual persons should not be added
```
Individual persons should not be given the permissions directly.

tuminoid replied:
Fixed.

kashifest commented on docs/releasing.md:
```diff
@@ -1,116 +1,141 @@
+`git clone git@github.com:metal3-io/cluster-api-provider-metal`
 
-### Process
+or if using existing repository, verify which remote the `metal3-io`
```
```suggestion
or if using existing repository, verify your intended remote is set  to `metal3-io`
```

tuminoid replied:
Fixed.

kashifest commented on docs/releasing.md:
```diff
@@ -1,116 +1,141 @@
+### Tags
 
-#### Automated Process
+First we create a main tag, that triggers release note creation and
```
main tag means ?

tuminoid replied:
Replaced with primary release tag, so its not confused with actual main branch.

kashifest commented on docs/releasing.md:
```diff
@@ -1,116 +1,141 @@
+image building processes.
 
-1. Create annotated tags:
+- Create an signed, annotated tag `git tag -s -a v1.x.y -m v1.x.y`
```
```suggestion
- Create a signed, annotated tag `git tag -s -a v1.x.y -m v1.x.y`
```

tuminoid replied:
Fixed.

kashifest commented on docs/releasing.md:
```diff
@@ -1,116 +1,141 @@
 
-1. Create annotated tags:
+- Create an signed, annotated tag `git tag -s -a v1.x.y -m v1.x.y`
+- Push all of the tags to the GitHub repository `git push origin v1.x.y`
```
```suggestion
- Push the tags to the GitHub repository `git push origin v1.x.y`
```

tuminoid replied:
Fixed.

kashifest commented on docs/releasing.md:
```diff
@@ -1,116 +1,141 @@
+- For any directory with `go.mod` in it (excluding `hack/tools`), create another
+  Git tag with directory prefix, ie. `git tag -s -a api/v1.x.y -m api/v1.x.y`.
+  In this repository, those directories are `api` and `tests`.
+  This is enables the tag to be used as a Go module version for any downstream
```
```suggestion
  This enables the tag to be used as a Go module version for any downstream
```

tuminoid replied:
Fixed.

kashifest commented on docs/releasing.md:
```diff
@@ -1,116 +1,141 @@
+  whatever release creation tagged to be manually checked.
+- For any superseded PRs (like CAPI uplifts, or commit revertions) that provide
+  no value to the release, create a summary line in `Superseded` section with
+  the PR ids and summary title. This way thye are acknowledged to be part of
```
```suggestion
  the PR ids and summary title. This way the changes are acknowledged to be part of
```

tuminoid replied:
Fixed

kashifest commented on docs/releasing.md:
```diff
@@ -1,116 +1,141 @@
+- For any superseded PRs (like CAPI uplifts, or commit revertions) that provide
+  no value to the release, create a summary line in `Superseded` section with
+  the PR ids and summary title. This way thye are acknowledged to be part of
+  the release, but not overwhelming the actual changes contained by the release.
```
```suggestion
  the release, but not overwhelming the important changes contained by the release.
```

tuminoid replied:
Fixed.

kashifest commented on docs/releasing.md:
```diff
@@ -1,116 +1,141 @@
 
-- Tag push access to the GitHub repository
-- GitHub Release creation access
+- Make sure Ironic and Mariadb version references are in place and of correct
```
```suggestion
- Make sure Ironic and Mariadb version references are in place with the correct
```

tuminoid replied:
Fixed.

kashifest commented on docs/releasing.md:
```diff
@@ -1,116 +1,141 @@
+[Ironic-image](https://github.com/metal3-io/ironic-image) and
+[Maridb-image](https://github.com/metal3-io/mariadb-image) repositories.
+Those repositories do not have separate releasing process, but for the sake of
+backwards compatibility, each of the CAPM3 releases is tagged in these
```
```suggestion
backwards compatibility, for each CAPM3 release we also create tags in these
```

tuminoid replied:
Fixed.

kashifest commented on docs/releasing.md:
```diff
@@ -1,116 +1,141 @@
-- GitHub Release creation access
+- Make sure Ironic and Mariadb version references are in place and of correct
+  version number.
+- If the release you're making isn't the actually latest release (new
```
```suggestion
- If the release you're making isn't the latest release (new
```

tuminoid replied:
Fixed.

kashifest commented on docs/releasing.md:
```diff
@@ -1,116 +1,141 @@
+- If the release you're making isn't the actually latest release (new
+  minor/major or new patch release in the latest minor), uncheck the
+  box for latest release.
+- If it is an release candidate, tick pre-release box.
```
```suggestion
- If it is a release candidate(rc) or pre-release , tick pre-release box.
```

tuminoid replied:
Fixed.

kashifest commented on docs/releasing.md:
```diff
@@ -119,28 +144,19 @@ For each minor or major release, two jobs need to be created :
+### Update testing matrix documentation
 
-### Create a new tag for Ironic image
+Update [testing matrix documentation](https://github.com/metal3-io/metal3-docs/pull/320)
```
This should probably point to a file upstream when the PR is merged.

furkatgofurov7 replied:
or the commit hash

tuminoid replied:
Yes, it'll point to real document link when it gets merged. For now, it points to PR.

furkatgofurov7 commented on docs/releasing.md:
```diff
@@ -1,116 +1,141 @@
-   folder
-1. Create a branch `release-0.x` for a minor release for backports and bug fixes.
-   This is not needed for patch releases.
+Before publishing CAPM3 release, we need to tag corresponding
```
Not necessarily before capm3 release I would say, but once CAPM3 is published (at least in the past we used to do so)

tuminoid replied:
Why would you want to do them later? When you go public with a release, why'd you intentionally want to have it half-ready? If you face issues tagging the images, or Quay wouldn't build them etc, then your release is broken.

furkatgofurov7 replied:
I would not recommend tagging it prior CAPM3 tag pushing, since the tag that is used as a prefix for these are the same as CAPM3 tag you will be pushing and you may end up in a situation where users might get confused with these (ironic tags available but not CAPM3, due to some technical reason when you could not push tag in CAPM3). So, IMO it is better to carry out these steps post CAPM3 tag pushing so we are backed up and referencing something that already exists upstream

tuminoid replied:
I'd still argue that before you press the publish button, nobody needs to care (won't even notice really) is more correct approach than publish first, and then scramble. Publishing a release sends out emails to everyone watching repo, might even trigger an automated workflow somewhere, so it is much bigger failure do it that way.

furkatgofurov7 commented on docs/releasing.md:
````diff
@@ -1,116 +1,141 @@
+```
+
+After this, [Ironic](https://quay.io/repository/metal3-io/ironic) and
+[Mariadb](https://quay.io/repository/metal3-io/mariadb) images will be built
````
```suggestion
[Mariadb](https://quay.io/repository/metal3-io/mariadb) container images will be built
```

tuminoid replied:
Fixed.

furkatgofurov7 commented on docs/releasing.md:
```diff
@@ -1,116 +1,141 @@
-   This is not needed for patch releases.
+Before publishing CAPM3 release, we need to tag corresponding
+[Ironic-image](https://github.com/metal3-io/ironic-image) and
+[Maridb-image](https://github.com/metal3-io/mariadb-image) repositories.
```
```suggestion
[Mariadb-image](https://github.com/metal3-io/mariadb-image) repositories.
```

tuminoid replied:
Fixed

furkatgofurov7 commented on docs/releasing.md:
```diff
@@ -1,116 +1,141 @@
+# Ironic
+git clone git@github.com:metal3-io/ironic-image.git
+cd ironic-image
+# if retagging old tag, add capm3-v1.x.y^{} to the line, e.g.
+# git tag -s -a capm3-v1.x.y -m capm3-v1.x.y capm3-v1.x.y^{}
+# this reuses the tag and resolves it into a sha for tagging
+git tag -s -a capm3-v1.x.y -m capm3-v1.x.y
+git push origin capm3-v1.x.y
```
I do not have an exact suggestion, but could we reuse this block later instead of repeating it and somehow say: `follow the exact same steps as above` ?

tuminoid replied:
I had it like that before, I'll change it back.

furkatgofurov7 commented on docs/releasing.md:
```diff
@@ -1,116 +1,141 @@
-- GitHub Release creation access
+- Make sure Ironic and Mariadb version references are in place and of correct
+  version number.
+- If the release you're making isn't the actually latest release (new
+  minor/major or new patch release in the latest minor), uncheck the
```
I think you meant: 

```suggestion
- If the release you're making isn't the actually latest release (new patch release in the latest-1 minor release series), uncheck the
```

tuminoid replied:
I don't get the point about `latest-1` ? 

Fixed the wording otherwise.

furkatgofurov7 replied:
The confusing part to me here is: if you release any major/minor release, that would always be the latest release and you do not need to uncheck the "latest release" button. 
But, if it is the patch release from the stable branch which is N-1 where N is latest stable branch (i.e you are publishing 1.3.5 patch release from release-1.3 but latest stable branch you have is release-1.4) then you need to uncheck it

tuminoid replied:
There is only one latest release per repo. In GitHub it means the release that is highlighted in the repository main page. 
The only time you tick the latest box is when you release from top off the latest release branch, otherwise must be unticked. Release from 1.2 branch will never be latest if you have 1.3 releases.

I'll see if there is better wording to make this clearer.

furkatgofurov7 commented on docs/releasing.md:
```diff
@@ -119,28 +144,19 @@ For each minor or major release, two jobs need to be created :
 Metal3-dev-env variables need to be modified. After a major or minor release,
-the new minor version (that follows CAPI versioning) should point to main for
-CAPM3 and the released version should point to the release branch.
+the new minor version should point to main for IPAM and the released version
```
why only IPAM and not CAPM3?

tuminoid replied:
Fixed.

furkatgofurov7 commented on docs/releasing.md:
```diff
@@ -119,28 +144,19 @@ For each minor or major release, two jobs need to be created :
 
-After this, Ironic container image will be created automatically with `capm3-v0.x.y` tag in Quay.
+We announce the release in Kubernetes slack in `#cluster-api-baremetal` channel
+and email `metal3-dev` list at Google groups.
```
```suggestion
and through the `metal3-dev` group mailing list.
```

tuminoid replied:
Fixed.

furkatgofurov7 reviewed (commented):
Thanks for the update PR. 
Few comments/suggestions from my side:

tuminoid commented on docs/releasing.md:
```diff
@@ -1,116 +1,141 @@
 
-### Expected artifacts
+These permissions are implicit for the org admins and repository admins.
+Release team gets their permissions via `metal3-release-team` membership. Adding
```
I clarified it is a GitHub team in the metal-3io organization. I don't have access to the organization page where its defined, nor will 99% of the readers. Admins will know where to look.

Fixed the wording.

tuminoid commented on docs/releasing.md:
```diff
@@ -1,116 +1,141 @@
+### Repository setup
 
-## Creating a release for CAPM3
+Clone out the repository:
```
Fixed

tuminoid commented on docs/releasing.md:
```diff
@@ -1,116 +1,141 @@
 
-There are two ways to do the release: automated (using GitHub workflow) or manual.
-The processes are as follows:
+- If creating a new minor branch, identify the commit your wish to create the
```
Fixed.

tuminoid commented on docs/releasing.md:
```diff
@@ -1,116 +1,141 @@
-The processes are as follows:
+- If creating a new minor branch, identify the commit your wish to create the
+  branch from, and create a branch `release-1.x`:
+  `git checkout <sha> -b release-1.x` and push it to remote:
+  `git push origin release-1.x` to create it
```
CLI is more exact, less prone to change, and easier to document. Rights required are exactly the same.

tuminoid commented on docs/releasing.md:
```diff
@@ -1,116 +1,141 @@
+image building processes.
 
-1. Create annotated tags:
+- Create an signed, annotated tag `git tag -s -a v1.x.y -m v1.x.y`
```
Fixed.

tuminoid commented on docs/releasing.md:
```diff
@@ -1,116 +1,141 @@
 
-1. Create annotated tags:
+- Create an signed, annotated tag `git tag -s -a v1.x.y -m v1.x.y`
+- Push all of the tags to the GitHub repository `git push origin v1.x.y`
```
Fixed.

tuminoid commented on docs/releasing.md:
````diff
@@ -1,116 +1,141 @@
-   git push origin v0.x.y
-   git push origin test/v0.x.y
-   ```
+We also need to do one or more tags for Go ecosystem:
````
This changes the intended meaning of the sentence. Subdirectories are discussed in the next sentence. 

Fixed the wording othewise.

tuminoid commented on docs/releasing.md:
```diff
@@ -1,116 +1,141 @@
+This triggers two things:
 
-1. Push the tags to the GitHub repository:
+- Release workflow on Github side that creates a draft release with correct
+  content, comparing the tag to previous tag
```
Fixed.

tuminoid commented on docs/releasing.md:
```diff
@@ -1,116 +1,141 @@
+repositories in the organization. Individual persons should not be added
+directly.
 
-1. A container image of the shared cluster-api-provider-metal3 manager
-1. A git tag
-1. A release on Github containing:
-     - A manifest file - `infrastructure-components.yaml`
-     - A metadata file - `metadata.yaml`
-     - A cluster template - `cluster-template.yaml`
-     - A file containing an example of variables to set - `example_variables.rc`
```
That is a good point. I added it back.

tuminoid commented:
Thanks @furkatgofurov7 @kashifest for spellcheck and catching couple missing things. Most of the things I've fixed, but some discussion left open. Let me know what you think.

/hold
There is also dependency to testing metrix PR, so I'll hold this so we get it merged first and can properly link to it.

furkatgofurov7 commented:
> Thanks @furkatgofurov7 @kashifest for spellcheck and catching couple missing things. Most of the things I've fixed, but some discussion left open. Let me know what you think.

Thank you, one more point I would like to bring up and forgot yesterday:
Can we left some notes in the doc somewhere (whereever you feel it fits, probably after pushing the tags sections) about the need we have to make sure that after tag is pushed, the image in quay has started actually building and the image is built successfully and it is available. 
The reason is, we have had a lot of technical issues in Quay when it comes to image building trigger (i.e trigger disabled, build unsuccessful) so person publishing a release has to make sure that quay has processed the tag and published the corresponding container images properly.

tuminoid commented:
> > Thanks @furkatgofurov7 @kashifest for spellcheck and catching couple missing things. Most of the things I've fixed, but some discussion left open. Let me know what you think.
> 
> Thank you, one more point I would like to bring up and forgot yesterday: Can we leave some notes in the doc somewhere (whereever you feel it fits, probably after pushing the tags sections) about the need we have to make sure that after tag is pushed, the image in quay has started actually building and the image is built successfully and it is available. The reason is, we have had a lot of technical issues in Quay when it comes to image building trigger (i.e trigger disabled, build unsuccessful) so person publishing a release has to make sure that quay has processed the tag and published the corresponding container images properly.
> 
> Edit: And ^ has to be followed for all container image building (CAPM3/IPAM/BMO/Ironic/mariadb)

I added the release artifacts and mention of successfully built container images there. BUT, as mentioned, we need to agree where to put all that stuff. CAPM3 != Metal3, and hence CAPM3 release process shouldn't contain Metal3 release process etc.

tuminoid commented:
There is two TODO left, one waiting for testing matrix PR to merge, and another for settling the proper location on the non-CAPM3 stuff. The latter needs input, so please share yours @furkatgofurov7 @kashifest .

furkatgofurov7 commented:
> BUT, as mentioned, we need to agree where to put all that stuff. CAPM3 != Metal3, and hence CAPM3 release process shouldn't contain Metal3 release process etc.

Partially agree, true but at the moment it is the repo, which owns kind of centralized status, and dependant on other repos (IPAM, BMO needs to be uplifted in it is go mods) and also main for other repos (whenever CAPM3 is release, you have to have a new ironic-image tag). 

> for settling the proper location on the non-CAPM3 stuff.

probably a new folder (i.e `release` in metal3-docs repo) ..

tuminoid commented:
> probably a new folder (i.e `release` in metal3-docs repo) ..

I've a WIP going https://github.com/metal3-io/metal3-docs/pull/321 for adding `releasing.md` under `processes` if you had missed it. It is very bare bones right now, but it seems we agree that all the Metal3 releasing, including dev-env modifications and Ironic/Mariadb tagging should be put there, until BMO has proper release process?

furkatgofurov7 commented on docs/releasing.md:
```diff
@@ -1,116 +1,158 @@
+  release.
+- Make sure Ironic and Mariadb version references are in place with the correct
+  version numbers.
+- If the release you're making isn't the latest release (new
+  minor or major, or new patch release in the latest minor release),
+  uncheck the box for latest release.
```
This is still confusing to me, though we discussed it earlier iterations, and also as you pointed out, when:

`you release from top off the latest release branch, otherwise must be unticked. Release from 1.2 branch will never be latest if you have 1.3 releases.`

But above, you are mention both new minor and major which should be ticked as latest when you publish one

tuminoid replied:
Fixed, better now?

furkatgofurov7 replied:
Thanks, it looks a lot better

furkatgofurov7 replied:
Feel free to resolve it

furkatgofurov7 reviewed (commented):
/lgtm cancel

One last comment that I think has to be addressed before I will add it back:

furkatgofurov7 commented:
/test-ubuntu-integration-main
/test-centos-e2e-integration-main

kashifest commented:
/lgtm
